### ISSUE TITLE
autodoc for generic classes should include the type parameters

### ISSUE THREAD

**Issue by jwodder** [sphinx-doc/sphinx#10568]
Currently, when autodoc renders a class of the form `class Foo(Generic[T])`, the resulting HTML just shows "class modulename.Foo", which omits information, as the type parameter is not shown.  The only way to show the type parameter at the moment is to use `:show-inheritance:` to add a line of the form "Bases: Generic[T]".  A generic class should instead be rendered as "class modulename.Foo[T]".

**Comment by AA-Turner** [comment]
PR welcome. What's the precedent here, though?

A

**Comment by jwodder** [comment]
I don't know of any precedent; I just feel that a class being generic should be indicated on the class name itself and not only as part of the base classes.

**Comment by tk0miya** [comment]
At present, autodoc and python domain expect to represent base classes via the `:Bases:` field. They also expect the parenthesis following the class name is a list of arguments list of class constructor.

I'm not sure the best way to represent the generic classes in the document. But we need to separate the class definition and constructor before the new implementation.

**Comment by arthur-tacca** [comment]
> What's the precedent here, though?

There's quite a bit of precedent. If you're talking about *usage* of generic types: I'm not sure what was possible in Python 3.5, when type hints and generics were first introduced, because I don't have easy access to an interpreter for it. But certainly by Python 3.6 (released in 2016) you could use type parameters and regular constructor parameters together with both built-in classes and user-defined classes based on `Generic[T]`. For example, 

    x = list[float]([1, 2, 3])
    y = MyGenericClass[int]("x")

The run-time types of the objects will be simply `list` and `MyGenericClass` (NOT `list[float]` and `MyGenericClass[int]`), but type hinters will take the static type of those variables to be `list[float]` and `MyGenericClass[int]`.

If you're talking about declaration rather than usage: that only came in with Python 3.12 (released in October 2023). Now instead of writing:

```python
T = TypeVar("T")
class MyGenericClass(SomeNonGenericBase, Generic[T]):
    ...
```

You can simply write (without declaring `T` beforehand):

```python
class MyGenericClass[T](SomeNonGenericBase):
    ...
```

Given the above, especially the usage syntax, it would make a lot of sense to me if you could automatically erase the `Generic[...]` base class from the list of base classes and show the generated docs for the class (regardless of how defined) as simply:

*class* **my_package.MyGenericClass**_[T](arg_1: str, arg_2: int)_

**Comment by arthur-tacca** [comment]
Just to preempt a potential implementation question: There's a slight complication with getting the type list when you have generic base classes.

In that case, there is no requirement to explicitly list the types with a Python 3.12-style type list or an older-style `Generic[...]` base. You can simply write the generic base classes you need. For example:

```python
T1 = TypeVar("T1")
T2 = TypeVar("T2")
T3 = TypeVar("T3")
class Base1(Generic[T1, T2]): pass
class Base2(Generic[T1, T2]): pass
class Derived(Base1[T1, T2], Base2[T2, T3]): pass
```

In that case, the type list is all the type variables used in the base list collected in order of first usage, i.e. for `Derived` it is `[T1, T2, T3]`.

If `Generic[T]` (or `Protocol[T]`) is listed as an explicit base class then it must include all type variables used (otherwise that is a mypy error). Presumably this also applies to Python 3.12-style class lists, though I haven't tried it. For example:

```python
class Derived(Base1[T1, T2], Base2[T2, T3], Generic[T3, T2, T1]): pass
```

In that case, the type list to `Generic[...]` instead defines the type list ordering. For example, it would be `[T3, T2, T1]` in the example above.

**Comment by picnixz** [comment]
I implemented PEP 695 since cpython asked for it but we'll probably rewrite it when 3.12 becomes the minimal version. However I did not consider any autodoc approach because extracting the type variables are painful depending on the python version...

I should have worked with astroid or implemented an improved version of the AST parser (+take vare of inheritance and different modules!!!). So at that time I just gave up I think.

**Comment by picnixz** [comment]
Yes, that's why I think we should postpone this until 3.12 becomes the base version. The main resaon is that the AST parser in <3.12 would not even recognize the syntax ! so I needed to make my own tokenizer (and also take into account some weird stuff because spaces are gobbled).

**Comment by arthur-tacca** [comment]
I had another look at this and it turns out you don't need to parse the code to pick out the type parameters of a `Generic[...]` base class. The trick is to use the `__orig_bases__` member, which, unlike `__bases__`, does not strip out the type parameters:

```python
>>> from typing import get_type_hints, TypeVar, Generic
>>> T = TypeVar("T")
>>> U = TypeVar("U")
>>> class C(Generic[T, U]): pass
...
>>> C.__bases__
(<class 'typing.Generic'>,)
>>> C.__bases__[0]
<class 'typing.Generic'>
>>> C.__orig_bases__
(typing.Generic[~T, ~U],)
>>> C.__orig_bases__[0]
typing.Generic[~T, ~U]
>>> C.__orig_bases__[0].__parameters__[0].__name__
'T'
```

**Comment by picnixz** [comment]
Actually, we are not handling the object directly. We are given a signature as a string. So I need to do a parser-like approach and cannot rely on runtime analysis (for instance, we support PEP 695 syntax given at the RST level directly, so there is no class object)

**Comment by arthur-tacca** [comment]
@picnixz I'm not very familiar with the autodoc code but tried to have a quick look at this. It does seem like the original class object is available in at least some situations – it's there in `self.object` of `ClassDocumenter`. In fact I found that I could just patch `ClassDocumenter.format_name()` (overriding the base class implementation `Documenter.format_name()`) to get the correct behaviour:

```python
    def format_name(self) -> str:
        """Format the name of *self.object*, with type parameters appended.
        """
        result = super().format_name()
        type_args = {}
        for p in getattr(self.object, "__type_params__", ()):
            type_args[p.__name__] = None
        for b in inspect.getorigbases(self.object) or ():
            if getattr(b, "__origin__", None) is typing.Generic:
                type_args.clear()
            for p in getattr(b, "__parameters__", ()):
                type_args[p.__name__] = None
        if type_args:
            result += "[" + ", ".join(type_args) + "]"
        return result
```

I'm not saying this is the right place put it in the code (e.g., it should probably not be in a function called "format_name"), and there should probably be an option to `autoclass` to say whether to do this, and I haven't included any code for supporting `*T` or `**T`. This is just meant to be a proof of concept that it's possible.

You talk about having the signature as a string. The only thing I could find like that in the code is if the signature is overridden by the user with a signature at the start of a docstring. For that, I think it's sensible to support PEP 695 type parameters (i.e. `class Foo[T](...)`, but that's a simple matter of the `ClassDocumenter` passing them unmodified to the output. The fancier work of the above function, converting parameters of base classes into PEP 695 type parameters, isn't relevant to that: if a user has explicitly said what the signature should look like, and (unlike when writing the actual Python code) they're not doing it that way because of old Python version compatibility, then it should be left alone anyway.

**Comment by picnixz** [comment]
Ah yes, you're right. The ClassDocumenter does have this. So... maybe I could make it work. But I may need some `if sys.version_info[:2] >= (3, 12)` (we have a single code for different versions) which I'm not very fond of. I don't have much time for this specific one so any help is welcomed (I can still review everything).

**Comment by arthur-tacca** [comment]
Thanks for your very fast reply (it is only 01:30 here 😃). The above code works fine on older Python versions with no version test needed. `getattr(self.object, "__type_params__", ())` simply returns the empty tuple in this case, and it seems using PEP 695 syntax in the output RST works fine because presumably core Sphinx is not relying on Python's own handling of that syntax.

**Comment by picnixz** [comment]
> it seems using PEP 695 syntax in the output RST works fine because presumably core Sphinx is not relying on Python's own handling of that syntax.

Yes, it's a different mechanism that I'd like to update once we are using Python 3.12 (but for now we still require Python 3.11).

> The above code works fine on older Python versions with no version test needed

Yes, but I don't know if we need more things to make everything work. Maybe not but since I'm too tired to check I'll leave to the tomorrow-me the task to check. Thanks for investigating it though!

**Comment by picnixz** [comment]
Ok, so in order not to give too much hope: I won't have time this week. So probably next week. I think I pretty much cleaned my backlog in CPython tasks so I can come back here more. I've got various tasks to do in Sphinx. But I'll try to make it before the end of the year.

**Comment by arthur-tacca** [comment]
@picnixz Many thanks for your hard work. BTW I updated the example function above to indicate what *T and **T might look like. I still imagine there are cases it doesn't cover (e.g., it wouldn't correctly qualify typevars from other modules because it just uses the bare `__name__`).

# Workaround for users

I've realised there's a simple workaround that works reasonably well if you have a small number of things to fix up (as my tiny library does). You can use the [`autodoc-process-signature` event](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#event-autodoc-process-signature) to add the type parameters to the start of the parameter list, like this (in your conf.py):

```python
def process_signature(app, what, name, obj, options, signature, return_annotation):
    if what == "class":
        if name == "aioresult.Future":
            signature = "[ResultT]" + (signature or "")
        elif name in ("aioresult.ResultBase", "aioresult.ResultCapture"):
            signature = "[ResultT_co]" + (signature or "")
    return signature, return_annotation

def setup(app):
    app.connect("autodoc-process-signature", process_signature)
```
